cur-test/cur/tests/stdlib/pattern-tree.rkt

```diff
@@ -0,0 +1,740 @@
+                                      ([true false => A]
+                                       [x true => B]))
+                                   #:env
+                                   (list (cons #'b #'Bool)
+                                         (cons #'a #'Bool)))
+              (pt-decl
+               #'a
+               (list
+                (pt-match
+                 #'true
+                 (pt-decl
+                  #'b
+                  (list
+                   (pt-match
+                    #'false
+                    (pt-body #'A))
+                   (pt-match
+                    #'true
+                    (pt-body #'B)))))
+                (pt-match
+                 #'x
+                 (pt-decl
+                  #'b
+                  (list
+                   (pt-match
+                    #'true
+                    (pt-body #'B)))))))))
+
+  ;; ADDITIONAL NESTING
+  (check-true
+   (pt-equal?
+    (create-pattern-tree #'((a b)
+                            ([z z => A]
+                             [(s (s (s b))) (s m) => b]))
+                         #:env
+                         (list
+                          (cons #'a #'Nat)
+                          (cons #'b #'Nat)))
+    (pt-decl
+     #'a
+     (list
+      (pt-match
+       #'z
+       (pt-decl
+        #'b
+        (list
+         (pt-match
+          #'z
+          (pt-body #'A)))))
+      (pt-match
+       #'(s temp28) ; note: for use in binding, we need to resolve temp bindings before positional input variables
+       (pt-decl     ; this means that we should effectively process temps by storing them on a stack, and processing in reverse
+        #'temp28
+        (list
+         (pt-match
+          #'(s temp29)
+          (pt-decl
+           #'temp29
+           (list
+            (pt-match
+             #'(s b30)
+             (pt-decl
+              #'b
+              (list
+               (pt-match
+                #'(s m31)
+                (pt-body #'b30))))))))))))))) ; note that we use the temp b not the match variable
+
+  (check-true
+   (pt-equal?
+    (create-pattern-tree
+     #'((e1 e2)
+        ([z z => A]
+         [(s (s e2)) z => B]
+         [(s (s e2)) (s m) => C]))
+     #:env
+     (list
+      (cons #'e1 #'Nat)
+      (cons #'e2 #'Nat)))
+    (pt-decl
+     #'e1
+     (list
+      (pt-match
+       #'z
+       (pt-decl
+        #'e2
+        (list
+         (pt-match
+          #'z
+          (pt-body #'A)))))
+      (pt-match
+       #'(s temp32)
+       (pt-decl
+        #'temp32
+        (list
+         (pt-match
+          #'(s e233)
+          (pt-decl
+           #'e2
+           (list
+            (pt-match
+             #'z
+             (pt-body #'B))
+            (pt-match
+             #'(s m34)
+             (pt-body #'C))))))))))))
+
+  ;; WITH AND WITHOUT TYPE CONTEXT
+  ; effectively, z = _ = n in this scenario
+  ; TODO PR103: Don't think this should pass as n is unbound.
```

Comment:
A possible argument for keeping this is that unbound identifiers do not necessarily need to be caught at this phase of re-write, given that the main purpose of the pattern-tree compiler is to restructure the input syntax

Reply:
I agree in principle, but it looks like much of the implementation is dependent on Cur specific APIs, and Cur is pretty sensitive about unbound identifiers.